File: tests/test_feature_wiring.py
# -*- coding: utf-8 -*-
"""
tests/test_feature_wiring.py - Ensures features are properly connected to UI

These tests catch cases where features exist in code but are not wired to the UI,
preventing "ghost features" that users can never access.

NOTE: These are static analysis tests that read source files directly.
They do NOT require the app to be running and don't use network fixtures.
"""

import unittest
import re
from pathlib import Path
import sys

# Skip conftest fixtures - these are pure static analysis tests
import pytest

pytestmark = pytest.mark.usefixtures()  # Override any auto-use fixtures

# Ensure project root is in path
sys.path.insert(0, str(Path(__file__).parent.parent))


class _TestFeatureWiringBase(unittest.TestCase):
    """Base methods for TestFeatureWiring."""

    def setUpClass(cls):
        """Load source files for analysis."""
        cls.root = Path(__file__).parent.parent
        cls.ui_components_path = cls.root / "ui_components.py"
        cls.zena_path = cls.root / "zena.py"
        cls.asgi_server_path = cls.root / "zena_mode" / "asgi_server.py"

        # Load file contents
        with open(cls.ui_components_path, "r", encoding="utf-8") as f:
            cls.ui_components_code = f.read()
        with open(cls.zena_path, "r", encoding="utf-8") as f:
            cls.zena_code = f.read()
        with open(cls.asgi_server_path, "r", encoding="utf-8") as f:
            cls.asgi_server_code = f.read()

    # =========================================================================
    # TUTORIAL SYSTEM
    # =========================================================================
    def test_tutorial_button_exists_in_ui(self):
        """Tutorial should have a trigger button in the UI."""
        # Check that start_tutorial or start_tour is called somewhere clickable
        self.assertIn(
            "start_tour", self.ui_components_code, "Tutorial function should be wired to a button in ui_components.py"
        )

    def test_tutorial_button_has_id(self):
        """Tutorial button should have an ID for testing."""
        self.assertIn(
            "ui-tour-btn", self.ui_components_code, "Tutorial button should have id='ui-tour-btn' for test automation"
        )

    # =========================================================================
    # HELP SYSTEM
    # =========================================================================
    def test_help_docs_indexed_on_startup(self):
        """Internal documentation should be indexed on app startup for help queries."""
        self.assertIn("index_internal_docs", self.zena_code, "index_internal_docs should be called in zena.py")
        # Check it's actually called, not just imported
        # Look for pattern: index_internal_docs(
        self.assertRegex(
            self.zena_code,
            r"index_internal_docs\s*\(",
            "index_internal_docs() should be called (not just imported) in zena.py",
        )

    # =========================================================================
    # VOICE LAB
    # =========================================================================
    def test_voice_lab_endpoint_exists(self):
        """Voice Lab should have a working endpoint in the ASGI server."""
        self.assertIn(
            "/voice/lab", self.asgi_server_code, "Voice Lab endpoint '/voice/lab' should exist in asgi_server.py"
        )

    def test_voice_lab_button_points_to_correct_url(self):
        """Voice Lab button should open the correct endpoint."""
        # Check that the iframe URL matches the endpoint
        self.assertIn(
            "localhost:8002/voice/lab",
            self.ui_components_code,
            "Voice Lab iframe should point to http://localhost:8002/voice/lab",
        )

    # =========================================================================
    # RAG DIALOG
    # =========================================================================
    def test_rag_dialog_exists(self):
        """RAG dialog should be defined and callable."""
        self.assertIn("setup_rag_dialog", self.ui_components_code, "setup_rag_dialog should exist in ui_components.py")


class TestFeatureWiring(_TestFeatureWiringBase, unittest.TestCase):
    """Tests that verify features are properly wired to the UI."""

    @classmethod

    # =========================================================================
    # MODEL MANAGEMENT
    # =========================================================================
    def test_model_switch_button_exists(self):
        """Model switching should be accessible from UI."""
        self.assertIn(
            "switch_to_model", self.ui_components_code, "switch_to_model function should exist for model switching"
        )

    # =========================================================================
    # SETTINGS DIALOG
    # =========================================================================
    def test_settings_dialog_wired(self):
        """Settings dialog should be wired to a button."""
        self.assertIn(
            "dialogs['settings'].open()", self.ui_components_code, "Settings dialog should be openable from UI"
        )

    # =========================================================================
    # QUALITY DASHBOARD / JUDGE
    # =========================================================================
    def test_quality_dashboard_exists(self):
        """Quality dashboard should be importable and used."""
        self.assertIn(
            "create_quality_tab", self.ui_components_code, "create_quality_tab should be called in the Judge dialog"
        )

    # =========================================================================
    # NEW CHAT BUTTON
    # =========================================================================
    def test_new_chat_button_exists(self):
        """New Chat button should exist with proper ID."""
        self.assertIn("BTN_NEW_CHAT", self.ui_components_code, "New Chat button with ID should exist in UI")

    # =========================================================================
    # ASGI API COMPLETENESS
    # =========================================================================
    def test_asgi_has_health_endpoint(self):
        """ASGI server should have /health endpoint."""
        self.assertIn("/health", self.asgi_server_code, "Health check endpoint should exist")

    def test_asgi_has_model_list_endpoint(self):
        """ASGI server should have model listing endpoint."""
        self.assertIn("/list", self.asgi_server_code, "Model list endpoint should exist")

    def test_asgi_has_tts_endpoint(self):
        """ASGI server should have TTS endpoint."""
        self.assertIn("/api/tts", self.asgi_server_code, "TTS endpoint should exist")

    def test_asgi_has_stt_endpoint(self):
        """ASGI server should have STT endpoint."""
        self.assertIn("/api/stt", self.asgi_server_code, "STT endpoint should exist")


class TestUIElementIDs(unittest.TestCase):
    """Tests that important UI elements have IDs for automation."""

    @classmethod
    def setUpClass(cls):
        """Setupclass."""
        cls.root = Path(__file__).parent.parent
        cls.registry_path = cls.root / "ui" / "registry.py"
        with open(cls.registry_path, "r", encoding="utf-8") as f:
            cls.registry_code = f.read()

    def test_critical_ui_ids_defined(self):
        """Critical UI elements should have IDs in the registry."""
        critical_ids = [
            "INPUT_CHAT",
            "BTN_SEND",
            "BTN_ATTACH",
            "BTN_VOICE",
            "BTN_SETTINGS",
            "BTN_NEW_CHAT",
        ]
        for id_name in critical_ids:
            with self.subTest(id_name=id_name):
                self.assertIn(id_name, self.registry_code, f"UI_IDS.{id_name} should be defined in registry.py")


class TestImportIntegrity(unittest.TestCase):
    """Tests that imported functions are actually used."""

    @classmethod
    def setUpClass(cls):
        cls.root = Path(__file__).parent.parent
        with open(cls.root / "zena.py", "r", encoding="utf-8") as f:
            cls.zena_code = f.read()

    def test_imported_tutorial_is_used(self):
        """start_tutorial should be used in UI components."""
        # Tutorial button is wired in ui_components.py, not zena.py
        ui_components_path = self.root / "ui_components.py"
        with open(ui_components_path, "r", encoding="utf-8") as f:
            ui_code = f.read()

        # Should have start_tutorial imported and called
        self.assertIn("start_tutorial", ui_code, "start_tutorial should be used in ui_components.py")

    def test_imported_help_system_is_used(self):
        """Imported index_internal_docs should be called."""
        import_match = re.search(r"from zena_mode\.help_system import index_internal_docs", self.zena_code)
        if import_match:
            after_import = self.zena_code[import_match.end() :]
            # Should have a function call
            self.assertRegex(
                after_import, r"index_internal_docs\s*\(", "index_internal_docs is imported but never called in zena.py"
            )


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: rust_output/src/test_feature_wiring.rs
/// tests/test_feature_wiring::py - Ensures features are properly connected to UI
/// 
/// These tests catch cases where features exist in code but are not wired to the UI,
/// preventing "ghost features" that users can never access.
/// 
/// NOTE: These are static analysis tests that read source files directly.
/// They do NOT require the app to be running and don't use network fixtures.

use anyhow::{Result, Context};
use regex::Regex;
use std::fs::File;
use std::io::{self, Read, Write};
use std::path::PathBuf;

pub static PYTESTMARK: std::sync::LazyLock<String /* pytest.mark.usefixtures */> = std::sync::LazyLock::new(|| Default::default());

/// Base methods for TestFeatureWiring.
#[derive(Debug, Clone)]
pub struct _TestFeatureWiringBase {
}

impl _TestFeatureWiringBase {
    /// Load source files for analysis.
    pub fn setUpClass(&self, cls: String) -> Result<()> {
        // Load source files for analysis.
        cls.root = PathBuf::from(file!()).parent().unwrap_or(std::path::Path::new("")).parent().unwrap_or(std::path::Path::new(""));
        cls.ui_components_path = (cls.root / "ui_components.py".to_string());
        cls.zena_path = (cls.root / "zena.py".to_string());
        cls.asgi_server_path = ((cls.root / "zena_mode".to_string()) / "asgi_server::py".to_string());
        let mut f = File::open(cls.ui_components_path)?;
        {
            cls.ui_components_code = f.read();
        }
        let mut f = File::open(cls.zena_path)?;
        {
            cls.zena_code = f.read();
        }
        let mut f = File::open(cls.asgi_server_path)?;
        {
            cls.asgi_server_code = f.read();
        }
    }
    /// Tutorial should have a trigger button in the UI.
    pub fn test_tutorial_button_exists_in_ui(&self) -> () {
        // Tutorial should have a trigger button in the UI.
        assert!(self.ui_components_code, "Tutorial function should be wired to a button in ui_components.py".to_string().contains("start_tour".to_string()));
    }
    /// Tutorial button should have an ID for testing.
    pub fn test_tutorial_button_has_id(&mut self) -> () {
        // Tutorial button should have an ID for testing.
        assert!(self.ui_components_code, "Tutorial button should have id='ui-tour-btn' for test automation".to_string().contains("ui-tour-btn".to_string()));
    }
    /// Internal documentation should be indexed on app startup for help queries.
    pub fn test_help_docs_indexed_on_startup(&self) -> () {
        // Internal documentation should be indexed on app startup for help queries.
        assert!(self.zena_code, "index_internal_docs should be called in zena.py".to_string().contains("index_internal_docs".to_string()));
        self.assertRegex(self.zena_code, "index_internal_docs\\s*\\(".to_string(), "index_internal_docs() should be called (not just imported) in zena.py".to_string());
    }
    /// Voice Lab should have a working endpoint in the ASGI server.
    pub fn test_voice_lab_endpoint_exists(&self) -> () {
        // Voice Lab should have a working endpoint in the ASGI server.
        assert!(self.asgi_server_code, "Voice Lab endpoint '/voice/lab' should exist in asgi_server::py".to_string().contains("/voice/lab".to_string()));
    }
    /// Voice Lab button should open the correct endpoint.
    pub fn test_voice_lab_button_points_to_correct_url(&self) -> () {
        // Voice Lab button should open the correct endpoint.
        assert!(self.ui_components_code, "Voice Lab iframe should point to http://localhost:8002/voice/lab".to_string().contains("localhost:8002/voice/lab".to_string()));
    }
    /// RAG dialog should be defined and callable.
    pub fn test_rag_dialog_exists(&self) -> () {
        // RAG dialog should be defined and callable.
        assert!(self.ui_components_code, "setup_rag_dialog should exist in ui_components.py".to_string().contains("setup_rag_dialog".to_string()));
    }
}

/// Tests that verify features are properly wired to the UI.
#[derive(Debug, Clone)]
pub struct TestFeatureWiring {
}

impl TestFeatureWiring {
    /// Model switching should be accessible from UI.
    pub fn test_model_switch_button_exists() -> () {
        // Model switching should be accessible from UI.
        assert!(self.ui_components_code, "switch_to_model function should exist for model switching".to_string().contains("switch_to_model".to_string()));
    }
    /// Settings dialog should be wired to a button.
    pub fn test_settings_dialog_wired(&self) -> Result<()> {
        // Settings dialog should be wired to a button.
        Ok(assert!(self.ui_components_code, "Settings dialog should be openable from UI".to_string().contains("dialogs['settings'].open()".to_string())))
    }
    /// Quality dashboard should be importable and used.
    pub fn test_quality_dashboard_exists(&self) -> () {
        // Quality dashboard should be importable and used.
        assert!(self.ui_components_code, "create_quality_tab should be called in the Judge dialog".to_string().contains("create_quality_tab".to_string()));
    }
    /// New Chat button should exist with proper ID.
    pub fn test_new_chat_button_exists(&self) -> () {
        // New Chat button should exist with proper ID.
        assert!(self.ui_components_code, "New Chat button with ID should exist in UI".to_string().contains("BTN_NEW_CHAT".to_string()));
    }
    /// ASGI server should have /health endpoint.
    pub fn test_asgi_has_health_endpoint(&self) -> () {
        // ASGI server should have /health endpoint.
        assert!(self.asgi_server_code, "Health check endpoint should exist".to_string().contains("/health".to_string()));
    }
    /// ASGI server should have model listing endpoint.
    pub fn test_asgi_has_model_list_endpoint(&self) -> () {
        // ASGI server should have model listing endpoint.
        assert!(self.asgi_server_code, "Model list endpoint should exist".to_string().contains("/list".to_string()));
    }
    /// ASGI server should have TTS endpoint.
    pub fn test_asgi_has_tts_endpoint(&self) -> () {
        // ASGI server should have TTS endpoint.
        assert!(self.asgi_server_code, "TTS endpoint should exist".to_string().contains("/api/tts".to_string()));
    }
    /// ASGI server should have STT endpoint.
    pub fn test_asgi_has_stt_endpoint(&self) -> () {
        // ASGI server should have STT endpoint.
        assert!(self.asgi_server_code, "STT endpoint should exist".to_string().contains("/api/stt".to_string()));
    }
}

/// Tests that important UI elements have IDs for automation.
#[derive(Debug, Clone)]
pub struct TestUIElementIDs {
}

impl TestUIElementIDs {
    /// Setupclass.
    pub fn setUpClass() -> Result<()> {
        // Setupclass.
        cls.root = PathBuf::from(file!()).parent().unwrap_or(std::path::Path::new("")).parent().unwrap_or(std::path::Path::new(""));
        cls.registry_path = ((cls.root / "ui".to_string()) / "registry::py".to_string());
        let mut f = File::open(cls.registry_path)?;
        {
            cls.registry_code = f.read();
        }
    }
    /// Critical UI elements should have IDs in the registry.
    pub fn test_critical_ui_ids_defined(&mut self) -> () {
        // Critical UI elements should have IDs in the registry.
        let mut critical_ids = vec!["INPUT_CHAT".to_string(), "BTN_SEND".to_string(), "BTN_ATTACH".to_string(), "BTN_VOICE".to_string(), "BTN_SETTINGS".to_string(), "BTN_NEW_CHAT".to_string()];
        for id_name in critical_ids.iter() {
            let _ctx = self.subTest(/* id_name= */ id_name);
            {
                assert!(self.registry_code, format!("UI_IDS.{} should be defined in registry::py", id_name).contains(id_name));
            }
        }
    }
}

/// Tests that imported functions are actually used.
#[derive(Debug, Clone)]
pub struct TestImportIntegrity {
}

impl TestImportIntegrity {
    pub fn setUpClass() -> Result<()> {
        cls.root = PathBuf::from(file!()).parent().unwrap_or(std::path::Path::new("")).parent().unwrap_or(std::path::Path::new(""));
        let mut f = File::open((cls.root / "zena.py".to_string()))?;
        {
            cls.zena_code = f.read();
        }
    }
    /// start_tutorial should be used in UI components.
    pub fn test_imported_tutorial_is_used(&mut self) -> Result<()> {
        // start_tutorial should be used in UI components.
        let mut ui_components_path = (self.root / "ui_components.py".to_string());
        let mut f = File::open(ui_components_path)?;
        {
            let mut ui_code = f.read();
        }
        Ok(assert!(ui_code, "start_tutorial should be used in ui_components.py".to_string().contains("start_tutorial".to_string())))
    }
    /// Imported index_internal_docs should be called.
    pub fn test_imported_help_system_is_used(&mut self) -> () {
        // Imported index_internal_docs should be called.
        let mut import_match = regex::Regex::new(&"from zena_mode\\.help_system import index_internal_docs".to_string()).unwrap().is_match(&self.zena_code);
        if import_match {
            let mut after_import = self.zena_code[import_match.end()..];
            self.assertRegex(after_import, "index_internal_docs\\s*\\(".to_string(), "index_internal_docs is imported but never called in zena.py".to_string());
        }
    }
}
